classes/services/PKPSubmissionService.php

```diff
@@ -0,0 +1,646 @@
+	 */
+	public function getWorkflowUrlByUserRoles($submission, $userId = null, $stageName = null) {
+
+		$request = Application::getRequest();
+
+		if (is_null($userId)) {
+			$user = $request->getUser();
+		} else {
+			$userDao = DAORegistry::getDAO('UserDAO');
+			$user = $userDao->getById($userId);
+		}
+
+		if (is_null($user)) {
+			return false;
+		}
+
+		$submissionContext = $request->getContext();
+
+		if (!$submissionContext || $submissionContext->getId() != $submission->getContextId()) {
+			$contextDao = Application::getContextDAO();
+			$submissionContext = $contextDao->getById($submission->getContextId());
+		}
+
+		$dispatcher = $request->getDispatcher();
+
+		// Check if the user is an author of this submission
+		$authorUserGroupIds = DAORegistry::getDAO('UserGroupDAO')->getUserGroupIdsByRoleId(ROLE_ID_AUTHOR);
+		$stageAssignmentsFactory = DAORegistry::getDAO('StageAssignmentDAO')->getBySubmissionAndStageId($submission->getId(), null, null, $user->getId());
+
+		$authorDashboard = false;
+		while ($stageAssignment = $stageAssignmentsFactory->next()) {
+			if (in_array($stageAssignment->getUserGroupId(), $authorUserGroupIds)) {
+				$authorDashboard = true;
+			}
+		}
+
+		// Send authors, journal managers and site admins to the submission
+		// wizard for incomplete submissions
+		if ($submission->getSubmissionProgress() > 0 &&
+				($authorDashboard || $user->hasRole(array(ROLE_ID_MANAGER, ROLE_ID_SITE_ADMIN), $submissionContext->getId()))) {
+			return $dispatcher->url(
+				$request,
+				ROUTE_PAGE,
+				$submissionContext->getPath(),
+				'submission',
+				'wizard',
+				$submission->getSubmissionProgress(),
+				array('submissionId' => $submission->getId())
+			);
+		}
+
+		// Send authors to author dashboard
+		if ($authorDashboard) {
+			return $dispatcher->url(
+				$request,
+				ROUTE_PAGE,
+				$submissionContext->getPath(),
+				'authorDashboard',
+				'submission',
+				$submission->getId()
+			);
+		}
+
+		// Send reviewers to review wizard
+		$reviewAssignment = DAORegistry::getDAO('ReviewAssignmentDAO')->getLastReviewRoundReviewAssignmentByReviewer($submission->getId(), $user->getId());
+		if ($reviewAssignment) {
+			return $dispatcher->url(
+				$request,
+				ROUTE_PAGE,
+				$submissionContext->getPath(),
+				'reviewer',
+				'submission',
+				$submission->getId()
+			);
+		}
+
+		// Give any other users the editorial workflow URL. If they can't access
+		// it, they'll be blocked there.
+		return $dispatcher->url(
+			$request,
+			ROUTE_PAGE,
+			$submissionContext->getPath(),
+			'workflow',
+			$stageName?$stageName:'access',
+			$submission->getId()
+		);
+	}
+
+	/**
+	 * Delete a submission
+	 *
+	 * @param $submissionId int
+	 */
+	public function deleteSubmission($id) {
+		Application::getSubmissionDAO()
+				->deleteById((int) $id);
+	}
+
+	/**
+	 * Check if a user can delete a submission
+	 *
+	 * @param $submission Submission|int Submission object or submission ID
```

Comment:
I'm not super fond of code that accepts mixed inputs like this -- I find it leads to bloat and risky edge cases. For example, did anyone check that ID to make sure it's in the right journal? Generally that happens when an ID is turned into an object but it's ambiguous who is responsible here. Generally I'd prefer we picked one behavior (e.g. take an ID or take a submission object) and stuck with it. But that's my preference.